rtshell/rtwatch.py

```diff
@@ -0,0 +1,228 @@
+    global counter
+    cevent = args[1]
+    if cevent == rtctree.component.Component.CFG_UPDATE_SET:
+        cevent = 'CFG_UPDATE_SET'
+    elif cevent == rtctree.component.Component.CFG_UPDATE_PARAM:
+        cevent = 'CFG_UPDATE_PARAM'
+    elif cevent == rtctree.component.Component.CFG_SET_SET:
+        cevent = 'CFG_SET_SET'
+    elif cevent == rtctree.component.Component.CFG_ADD_SET:
+        cevent = 'CFG_ADD_SET'
+    elif cevent == rtctree.component.Component.CFG_REMOVE_SET:
+        cevent = 'CFG_REMOVE_SET'
+    elif cevent == rtctree.component.Component.CFG_ACTIVATE_SET:
+        cevent = 'CFG_ACTIVATE_SET'
+    print('[{0}] {1} {2}'.format(time.time(), args[0], cevent))
+    counter += 1
+
+def heartbeat_cb(event, args, args2):
+    global counter
+    print('[{0}] {1}'.format(args[1], args[0]))
+    counter += 1
+
+def fsm_event_cb(event, args, args2):
+    global counter
+    print('[{0}] {1} {2}'.format(time.time(), args[0], args[1]))
+    counter += 1
+
+filtermap = {
+    'RTC_STATUS': ('rtc_status', rtc_status_cb),
+    'COMPONENT_PROFILE': ('component_profile', component_profile_cb),
+    'EC_EVENT': ('ec_event', ec_event_cb),
+    'PORT_EVENT': ('port_event', port_event_cb),
+    'CONFIG_EVENT': ('config_event', config_event_cb),
+    'FSM_EVENT': ('fsm_event', fsm_event_cb),
+    'HEARTBEAT': ('heartbeat', heartbeat_cb),
+}
+
+def clean_events(rtcs):
+    # Remove all the event hooks that were added
+    for i in rtcs:
+        i.dynamic = False
+
+def print_logs(paths, options, tree=None):
+    global counter
+    for p in paths:
+        path, port = rtctree.path.parse_path(p[1])
+        if port:
+            raise rts_exceptions.NotAComponentError(p[0])
+        if not path[-1]:
+            raise rts_exceptions.NotAComponentError(p[0])
+        p.append(path)
+
+    if not tree:
+        parsed = [p[2] for p in paths]
+        tree = rtctree.tree.RTCTree(paths=parsed, filter=parsed)
+
+    rtcs = []
+    for p in paths:
+        if not tree.has_path(p[2]):
+            raise rts_exceptions.NoSuchObjectError(p[0])
+        rtc = tree.get_node(p[2])
+        if rtc.is_zombie:
+            raise rts_exceptions.ZombieObjectError(p[0])
+        if not rtc.is_component:
+            raise rts_exceptions.NotAComponentError(p[0])
+
+        rtc.dynamic = True
+        if len(options.filters) == 0 or 'ALL' in options.filters:
+            for (k,v) in filtermap.items():
+                rtc.add_callback(v[0], v[1])
+        else:
+            for f in options.filters:
+                try:
+                    v = filtermap[f]
+                    rtc.add_callback(v[0], v[1])
+                except KeyError:
+                    print('Unknown filter: {0}'.format(f))
+        rtcs.append(rtc)
+
+    # Wait for a keyboard interrupt
+    counter = 0
+    atexit.register(clean_events, rtcs)
+    while True:
+        if options.number > 0 and counter >= options.number:
+            break
+        time.sleep(0.5)
```

Comment:
Instead of a busy loop, this function should wait on an `Event` object. The callbacks can check the counter for passing the limit and signal the event when it does, waking up this main thread and letting it exit.